### PR TITLE
Enable epochs handler integration part 7

### PR DESCRIPTION
**Pull request by sstanculeanu**
- added new method SwitchJailWaitingEnableEpoch
- completely use enableEpochsHandler into rewards system
- new stub specific for nodes coordinator to avoid cyclic dependencies
- cleanup
- added back updateEpochFlags on indexHashedNodesCoordinator missed in part 6 https://github.com/ElrondNetwork/elrond-go/pull/4158

**Notes:**
1. In order to properly check if the old flags match the new methods, one who review should check old conditions from removed `EpochConfirmed` method, the value from config that was passed before, and the line from `EpochConfirmed` from `enableEpochsHandler.go` which uses the same config value.
2. No focus on long tests so far, they will be fixed on the last PR once all components use `enableEpochsHandler`.

### CODE REVIEW

File: epochStart/metachain/rewards.go [hunk 1]
```diff
@@ -59,7 +59,7 @@ func (rc *rewardsCreator) CreateRewardsMiniBlocks(
 	defer rc.mutRewardsData.Unlock()
 
 	rc.clean()
-	rc.flagDelegationSystemSCEnabled.SetValue(metaBlock.GetEpoch() >= rc.delegationSystemSCEnableEpoch)
+	rc.flagDelegationSystemSCEnabled.SetValue(metaBlock.GetEpoch() >= rc.enableEpochsHandler.StakingV2EnableEpoch())
```
Comment (bogdan-rosianu) [inline epochStart/metachain/rewards.go:62]: why do we still need this flag?

File: epochStart/metachain/rewardsCreatorProxy.go [hunk 2]
```diff
@@ -135,7 +130,7 @@ func (rcp *rewardsCreatorProxy) changeRewardCreatorIfNeeded(epoch uint32) error
 	rcp.mutRc.Lock()
 	defer rcp.mutRc.Unlock()
 
-	if epoch > rcp.epochEnableV2 {
+	if epoch > rcp.args.EnableEpochsHandler.StakingV2EnableEpoch() {
```
Comment (bogdan-rosianu) [inline epochStart/metachain/rewardsCreatorProxy.go:133]: not 100% sure, but you could have used `IsStakingV2FlagEnabledForActivationEpochCompleted`

File: epochStart/metachain/rewardsV2.go [hunk 3]
```diff
@@ -99,7 +99,7 @@ func (rc *rewardsCreatorV2) CreateRewardsMiniBlocks(
 
 	miniBlocks := rc.initializeRewardsMiniBlocks()
 	rc.clean()
-	rc.flagDelegationSystemSCEnabled.SetValue(metaBlock.GetEpoch() >= rc.delegationSystemSCEnableEpoch)
+	rc.flagDelegationSystemSCEnabled.SetValue(metaBlock.GetEpoch() >= rc.enableEpochsHandler.StakingV2EnableEpoch())
```
Comment (bogdan-rosianu) [inline epochStart/metachain/rewardsV2.go:102]: still need the flag?

File: sharding/mock/enableEpochsHandlerStub.go [hunk 4]
```diff
@@ -0,0 +1,394 @@
+package mock
+
+type EnableEpochsHandlerStub struct {
+	WaitingListFixEnableEpochField uint32
+}
+
+// BlockGasAndFeesReCheckEnableEpoch returns 0
+func (stub *EnableEpochsHandlerStub) BlockGasAndFeesReCheckEnableEpoch() uint32 {
```
Comment (bogdan-rosianu) [inline sharding/mock/enableEpochsHandlerStub.go:8]: this is not a stub, but rather a mock, as you always return 0. also, don't we have this stub somewhere else?

File: sharding/mock/enableEpochsHandlerStub.go [hunk 5]
```diff
@@ -0,0 +1,394 @@
+package mock
+
+type EnableEpochsHandlerStub struct {
```
Comment (bogdan-rosianu) [inline sharding/mock/enableEpochsHandlerStub.go:3]: missing comment

File: sharding/nodesCoordinator/hashValidatorShuffler.go [hunk 6]
```diff
@@ -69,27 +69,25 @@ func NewHashValidatorsShuffler(args *NodesShufflerArgs) (*randHashShuffler, erro
 	if args == nil {
 		return nil, ErrNilNodeShufflerArguments
 	}
+	if check.IfNil(args.EnableEpochsHandler) {
+		return nil, ErrNilEnableEpochsHandler
+	}
 
 	var configs []config.MaxNodesChangeConfig
 
 	log.Debug("hashValidatorShuffler: enable epoch for max nodes change", "epoch", args.MaxNodesEnableConfig)
```
Comment (bogdan-rosianu) [inline sharding/nodesCoordinator/hashValidatorShuffler.go:78]: forgotten flag?

File: epochStart/metachain/rewards.go (hunk 1 above)
Comment (sstanculeanu) [reply epochStart/metachain/rewards.go:62]: kept the flag as it is set based on the epoch from metablock, not the epoch received on `EpochConfirmed`

File: sharding/mock/enableEpochsHandlerStub.go (hunk 4 above)
Comment (sstanculeanu) [reply sharding/mock/enableEpochsHandlerStub.go:8]: renamed it to `enableEpochsHandlerStub`..indeed we do have the stub in `testscommon`, but using it into `nodesCoordinator` package was leading to cyclic imports.

I decided that creating a new mock, with the needed functionality for this package was an easier change than moving all the tests from package `nodesCoordinator` to a new `nodesCoordinator_test` package, as they directly use a lot of unexported methods

File: sharding/nodesCoordinator/hashValidatorShuffler.go (hunk 6 above)
Comment (sstanculeanu) [reply sharding/nodesCoordinator/hashValidatorShuffler.go:78]: `MaxNodesEnableConfig` is a config, not a specific epoch from `enableEpochs.toml`

File: epochStart/metachain/rewardsCreatorProxy.go (hunk 2 above)
Comment (sstanculeanu) [reply epochStart/metachain/rewardsCreatorProxy.go:133]: same as above, the `epoch` param comes from metablock

File: epochStart/metachain/rewardsV2.go (hunk 3 above)
Comment (sstanculeanu) [reply epochStart/metachain/rewardsV2.go:102]: same here, epoch is from metablock

**Review by bogdan-rosianu — APPROVED** [review]

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go [hunk 7]
```diff
@@ -1186,3 +1193,8 @@ func createValidatorInfoFromBody(
 
 	return allValidatorInfo, nil
 }
+
+func (ihnc *indexHashedNodesCoordinator) updateEpochFlags(epoch uint32) {
+	ihnc.flagWaitingListFix.SetValue(epoch >= ihnc.enableEpochsHandler.WaitingListFixEnableEpoch())
+	log.Debug("indexHashedNodesCoordinator: waiting list fix", "enabled", ihnc.flagWaitingListFix.IsSet())
+}
```
Comment (ssd04) [inline sharding/nodesCoordinator/indexHashedNodesCoordinator.go:1200]: :+1:

**Pull request merged**